### DISCUSSION TITLE
LilyGO-T-RELAY 4 channel ESP32

### DISCUSSION THREAD

**Discussion by dac2020** [arendst/Tasmota#16054] [Show and tell]
For anyone who's playing with this board, it was nearly impossible to find information on this so I thought I should share.

I accidentally ordered the above board without the U2T kit adapter to flash it. I'm wanting to get something done over the weekend and am not sure I'll be using this board permanently as I learn about tasmota so I decide to try using serial. 

From what I've read, this board has three serial ports. 
A quick look here seems to only show two sets of R/X pins. 
https://tasmota.github.io/docs/Pinouts/#esp32-wrover

This seems to show pins 17 (rx) and 18 (tx) are what I'll need or alternatively, 34/35.
The image in the link shows gpio1 tx, gpio3 rx and gpio9 rx, gpio10 tx.

However, the board doesn't have any of those, it only shows IOxx pins and none are the above.
It could be pins IO34 and IO35 so I tried that but I get nothing.

I cannot find any information on what pins I need to connect to get serial access.

I have the esp32 module on this. ESP32-WROVER-B but as mentioned, that link being shared, I don't have those pins in the header.
The header labels are all IOxx and none of the numbers are pins I've found anywhere in this link or on the net.
The top pins on the right show 34 and 35 and I have IO34 and IO35 on the header label but those didn't work.

I took a nice and thick C-cable and carefully cut it open to reveal only four wires. 
Two power, two data I guess since two of the wires are obviously thicker than the others so must be power.

It's not a full cable but I hoped it meant it would simply have power and data connectors but so far, I cannot get much from a serial connection, just a lot of garbage at all baud rates and flow control settings. Definitely what seems to be a stream of data as it corresponds with the initial booting and testing of each relay then a pulse.

Update, ah, finally got it.
115200, 8N1, XON/XOFF.
Must have rx, tx AND GND connected. Without ground, it goes crazy with data on the terminal screen.
`
..ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x12 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0018,len:4
load:0x3fff001c,len:1044
load:0x40078000,len:10124
load:0x40080400,len:5856
entry 0x400806a8
setup
scan start
scan done
6 networks found
1: mylan (-62)*
2: neighbor1 (-89)*
3: neighbor2 (-89)*
4: neighbor3 (-89)*
5: neighbor4 (-89)*
6: neighbor5 (-90) 

Connecting to xinyuandianzi
.....................
`

Interesting that it sees the local wilreless networks. 
Says it's connecting to 'xinyuandianzi'. No idea what that is, there's no wifi around here with that name.

Anyhow, now I need to find information that explains how to load tasmota onto this and I finally get to see it in action then I'll play with a broker I've installed on a local ubuntu server.

**Comment by sfromis** [comment]
There is only one relevant serial port for flashing. TX/RX on gpio 1/3. Yes, that's not broken out in headers, but that's what you get for not buying the standard USB/serial adapter sold by Lilygo. And you need the usual power+gnd, and grounding gpio 0 to get in flashing mode.

**Reply by dac2020** [reply]
I already said in my other post that I didn't realize at the time that this board didn't come with the adapter but either way, I wasn't able to find it on Amazon using the model/parts they share.

>And you need the usual power+gnd, and grounding gpio 0 to get in flashing mode.

I don't know what any of that means. I'm too new to this and figuring it out with your input and searching on the net :).

I *assume* it means I have to short some pins to put into a receiving mode so I can send it tasmota?

**Reply by sfromis** [reply]
Read https://tasmota.github.io/docs/Getting-Started/
While that refers to ESP8266, the "classic" ESP32 chips are not different when it comes to establishing wires for flashing.

As for the other thread, controlling flashing uses at least 5 wires, maybe 6, thus the "fake USB" connector has more than 4.

**Reply by dac2020** [reply]
I'm not quite following on the 5 wires thing. There are 5 wires connected, the two power leads to the power input and the rx/tx/gnd via the usb-c connector. Do you mean I still cannot flash this yet? I have serial access, isn't that all I need?

**Reply by sfromis** [reply]
5 wires is explained in the guide I linked, what you minimum need is power, ground, tx, rx and gpio0. Not sure what you mean by "two power leads", as the only power required for flashing is that the ESP chip is powered with 3.3V.

Do you main mains power? Do **not** have that connected while flashing, doing that can cost you a lot ...

**Reply by dac2020** [reply]
No mains at all, just the serial connection to the usb-c connector and 12VDC powering the board using the power input terminals. You can see the 12V input on the 4 channel board image here.

https://github.com/Xinyuan-LilyGO/LilyGo-T-Relay

Do you mean the board needs to be powered from the same usb-c then, using the left over power wires I also have but not using. 

Pretty sure this is ready to flash but the doc says short gnd and gpio00 and of course, I don't see any gpio00.
So, have to dig some more first.

**Reply by sfromis** [reply]
While it may be possible to have it powered via 12V, my preferred setup, for simplicity, would still be to have no other power source while flashing, unless it does not work that way.

You do very much see a gpio0 if you look at the specs for the ESP32-WROVER module. Yes, it is not broken out, like the tx/rx you need are not either. I'd still expect the gpio0 pin to be connected to the fake USB socket, along with tx/rx, meaning that at least 5 pins should be connected there.

As I mentioned previously, the board is really not "friendly" to flashing without their adapter dongle.

**Reply by dac2020** [reply]
Yes, I see all the GPIO's but my point is they aren't all tied to the header and even that are labelled IO so not 100% sure those mean GPIO. I assume they do.

Yes, it's not friendly to flash for sure and the frustrating thing is that I'm doing this just to get some hands on experience doing it as I've ordered one or two devices that already have tasmota installed :).

I just wanted to play with this over the weekend when I have a few minutes here, there.

I can't find anything that gives any info on which lead might be GPIO00 so I guess this is as far as I can get.

**Reply by dac2020** [reply]
I guess I could try soldering a lead to GPIO00. Can I use the GND from the usb-c connector then?

**Reply by sfromis** [reply]
Yes. You could also hand-hold a connection while powering up.

> Yes, I see all the GPIO's

If you look at the ESP32-WROOM module pinout, that's where you see all the gpios, and not all of them are tied to the header, by far not.

To find info on what's connected, use your multimeter to check where it beeps.

**Reply by Jason2866** [reply]
Safe working way is to solder directly the needed wires to the wrover module tx rx gpio0 3.3V gnd to the USB serial device (preferred the "golden" ch340). Powering the wrover module from there.
**I recommend never to buy such crippeled stuff,** LilyGO tries to sell there very very overprized usb serial adapter and did everything to make it complicate as much as possible to flash the device the usual way.
There is no reason for NOT breaking out the needed gpios to a header. LilyGO is a company fooling there buyers. I bought no device from them since the introduced this shit.
A good [alternative](https://templates.blakadder.com/robotdyn_ESP32R4.html) for example

**Reply by dac2020** [reply]
>You could also hand-hold a connection while powering up.

Yes, I thought about that after I walked away from my reply. I'll give that a try shortly :).

I'm brand new to trying tasmota and part of the learning is what hardware to buy for sure. They do make it weird for someone using their board don't they. There are some posts on their Amazon page of others who seem to have flashed it but they likely used the adapter which I thought was included when I ordered. 

No big deal, I'll try touching the lead and see what happens.

**Comment by dac2020** [comment]
Well, the board smoked when I applied power via the usb-c so I'm done for now.
I guess this whole post should be removed because I will not be buying another of these boards.

I did buy another board at the same time because I see reviews saying some flashed it. 
https://www.amazon.com/dp/B071KFX63R

The chip is PSF-B so I'll look at the page that was shared. Was so hoping to get to try tasmota over this weekend.

Update: And of course, none of the leads for headers are marked to show which connect to rx/tx.

**Reply by Jason2866** [reply]
Sorry to hear you wasted money for buying this user cheating board.
(Re)using a USB-C connector in a board design which can not handle plugin a 5V USB-C power supply is...

**Reply by dac2020** [reply]
is?

**Reply by Jason2866** [reply]
> is?

i try not to use very agressive or swear words ;-)
(Sorry for my editing errors)

**Reply by dac2020** [reply]
>@Jason2866
>Jason2866
>42 minutes ago
>Collaborator
>Sorry to hear you wasted money for buying this user cheating board.
>(Re)using a USB-C connector in a board design which can not handle plugin a 5V USB-C power supply is...

I still don't know what the end of your comment means. I'm not easily offended by swear words :).

**Reply by sfromis** [reply]
Just imagine the swear words yourself. If you avoid the softer ones, you're probably close :stuck_out_tongue_closed_eyes:

**Reply by sfromis** [reply]
Schematic here, if you're still interested....
https://github.com/Xinyuan-LilyGO/LilyGo-T-Relay/tree/main/Schematic

**Reply by dac2020** [reply]
No idea what you are talking about. I assume it's something to do with encouraging a 70 year old man to learn because this seems like an interesting project that members would love to see being used by many. And not to scare those new people off by being rude or mean to them. I'm assuming all that :).

The link you show I had shared already mentioning that none of the header labels match the chip. I even used a magnifying glass to try and see if I could see the leads but it seems to be a layered board.

**Reply by Jason2866** [reply]
Oh sorry, the comment was not ment in your direction. Everything goes in direction of LilyGo!!
You had a bad user experince since LilyGo made a board which is against all common norms and introduced a *#/* own "standard" for flashing and did not make the needed GPIOs available on the PCB for flashing.
The only reason i see for doing this, is to sell there highly priced USB serial adapter.

**Reply by Jason2866** [reply]
That was, was i ment with the `...`

**Reply by Jason2866** [reply]
70 years **young!** I hope my brain will work so well as yours when doing new stuff.
I think i can speak for all people around Tasmota we do find it great that all ages are trying Tasmota.
The youngest/oldest i know where under <10 years and >80 years. So you are a young Tasmota user :-)
Dont hesitate to join our [Tasmota Discord Server](https://discord.gg/Ks2Kzd4). Yes more new fun stuff to learn. It is the place where Tasmota users met to talk about Tasmota. Everyone is welcome, beginers and advanced Tasmotians

**Reply by dac2020** [reply]
I appreciate that it wasn't against my lack of knowledge. I was excited so bought what I could at the moment and will find better devices. 
I wanted to start with automatic a sprinkler system and another thing so thought I would start with the delay boards.

Life is all about learning and I'm assuming I will blow up more boards but that's ok. The end result of being able to use an interesting project like this will make it worth the time and cost of making mistakes :).

**Comment by yoannchaudet** [comment]
If that helps anyone else, I went through the ~trouble~ task of flashing the 8-relay version of this board with Tasmota.

Happy to report that it was easier than I feared (although I do have the adapter to make the process simple)! I documented the steps I followed in https://github.com/yoannchaudet/LilyGo-T-Relay-Tasmota.

**Reply by sfromis** [reply]
Well, using the adapter it was indeed expected to be just as easy as flashing other devices with more friendly designs. It was just a crap move to design and sell a device outright hostile to flashing without having to buy an overpriced adapter on the side, when it would be no real cost to break out the pins in an accessible way, as done with nearly all other boards. I've even got a Lilygo device meant to be flashed with a special adapter, but still having the relevant pins broken out for easy access.

**Reply by barbudor** [reply]
Thanks @yoannchaudet for documenting
If there is not a  template page on templates.blakadder.com could you please add one
In both case, having a link to your guide would be useful for the future

**Reply by yoannchaudet** [reply]
I submitted the 8-relay template for review yesterday ❤️

**Reply by fb-pilot** [reply]
just read here about the boards ... I found a link to more information including a schematic at AliExpress :
https://github.com/Xinyuan-LilyGO/LilyGo-T-Relay

**Reply by dexterbot80** [reply]
Very nicely documented, let's now connect a shunt from Juncktek to RS485 and send the data further on CAN.
This would be the icing on the cake :)

**Comment by fapoo** [comment]
Hi guys, I ordered the corresponding 8-channel relay last week and received it yesterday. Apparently you don't need the T-U2T adapter anymore since RX and TX are now pins on the board (picture). 
Lucky me, because I did not order the adapter but now I managed to flash it with tasmota. When running over usb-c (5V) the board draws 30-50 mA with all relays of, but about 600 mA when all switched on.
![IMG_7657](https://github.com/arendst/Tasmota/assets/45537914/df4a0242-de05-4733-9bd4-f8bfeec50778)

**Reply by Noschvie** [reply]
Looks nice!
